### PR TITLE
Add typing to ``BaseChecker`` ``msgs`` and ``reports``

### PR DESCRIPTION
**Pull request by DanielNoord**
- [x] Add yourself to CONTRIBUTORS if you are a new contributor.
- [x] Write a good description on what the PR does.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :hammer: Refactoring   |

## Description

### CODE REVIEW

File: pylint/checkers/classes/class_checker.py [hunk 1]
```diff
@@ -428,7 +429,9 @@ def _has_same_layout_slots(slots, assigned_value):
     return False
 
 
-MSGS = {  # pylint: disable=consider-using-namedtuple-or-dataclass
+MSGS: dict[
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/classes/class_checker.py:432]: I'd rather put that directly in the checker as ``msgs`` so the typing is implicit, what do you think ?

File: pylint/typing.py [hunk 2]
```diff
@@ -117,3 +121,14 @@ class ManagedMessage(NamedTuple):
 CheckerT_co = TypeVar("CheckerT_co", bound="BaseChecker", covariant=True)
 AstCallbackMethod = Callable[[CheckerT_co, "nodes.NodeNG"], None]
 """Callable representing a visit or leave method."""
+
+ReportsCallable = Callable[["Section", "LinterStats", Optional["LinterStats"]], None]
+"""Callable to create a report."""
+
```
Comment (Pierre-Sassoulas) [inline pylint/typing.py:127]: Thinking out loud here, but what if we introduce a class that does all this and make it compatible with the legacy ? (I.e. we handle both for a while but encourage the use of the class)

```python
class RawMessageDefinition(NamedTuple):

   """ This class look a lot like a ``MessageDefinition`` but without having to give the checker. """

   msgid: str #Or another class taking into account that it's a special string with a letter and a number
   symbol: str #Or another class taking into account that it's a special string with only letters and '-'
   ...
```

File: pylint/checkers/classes/class_checker.py (hunk 1 above)
Comment (DanielNoord) [reply pylint/checkers/classes/class_checker.py:432]: That doesn't work as dictionaries are invariant. See https://mypy.readthedocs.io/en/stable/common_issues.html#variance

File: pylint/typing.py (hunk 2 above)
Comment (DanielNoord) [reply pylint/typing.py:127]: Probably need to this to complete the typing of `BaseChecker` as it uses `**definition` a couple of times.

For now, I think this is fine though. Small steps towards a complete and fully typed version.

**Review by Pierre-Sassoulas — APPROVED** [review]

**Pull request merged**